swarm/src/behaviour/deny_list.rs

```diff
@@ -0,0 +1,60 @@
+use crate::behaviour::THandlerInEvent;
+use crate::{dummy, CloseConnection, NetworkBehaviour, NetworkBehaviourAction, PollParameters};
+use libp2p_core::{ConnectedPoint, PeerId};
+use std::collections::{HashSet, VecDeque};
+use std::task::{Context, Poll};
+
+#[derive(Default)]
+pub struct Behaviour {
+    list: HashSet<PeerId>,
+    to_disconnect: VecDeque<PeerId>,
+}
+
+impl Behaviour {
+    pub fn add_peer(&mut self, peer: PeerId) {
+        self.list.insert(peer);
```

Comment:
```suggestion
pub struct Behaviour<T = ()> {
    list: HashMap<PeerId, T>,
    to_disconnect: VecDeque<PeerId>,
}

impl Behaviour<()> {
    pub fn add_peer(&mut self, peer: PeerId) {
        self.list.insert(peer, ());
        self.to_disconnect.push_back(peer);
    }
}

impl<T> Behaviour<T> {
    pub fn add_peer_because(&mut self, peer: PeerId, reason: T) {
        self.list.insert(peer, reason);
```

We could allow an optional `reason` parameter that we'd then also include in the `Denied` struct.  Might be useful if there are multiple reasons why a peer can be banned.
If we'd additionally expose what peers are currently banned (and for what reason) it would allow a more sophisticated management of banned peers. E.g. if a peer was only banned because an Application-Layer limit was reached.

But not sure if it's worth the extra complexity. Could wait until there is actually a use-case for this.